# Image Classification - The Multi-class Weather Dataset

**Submission deadline: Friday 5 April, 11:55pm**

**Assessment weight: 15% of the total unit assessment.**

**Versions**

- Wednesday 13 March: Initial release

*Unless a Special Consideration request has been submitted and approved, a 5% penalty (of the total possible mark of the task) will be applied for each day a written report or presentation assessment is not submitted, up until the 7th day (including weekends). After the 7th day, a grade of ‘0’ will be awarded even if the assessment is submitted. The submission time for all uploaded assessments is **11:55 pm**. A 1-hour grace period will be provided to students who experience a technical concern. For any late submission of time-sensitive tasks, such as scheduled tests/exams, performance assessments/presentations, and/or scheduled practical assessments/labs, please apply for [Special Consideration](https://students.mq.edu.au/study/assessment-exams/special-consideration).*

In this assignment you will complete tasks for an end-to-end image classification application. We will train and test the data using the Multi-class Weather Dataset (MWD):

- https://data.mendeley.com/datasets/4drtyfjtfy/1

The MWD contains labelled images representing various weather scenarios. It is a small and popular dataset for practice with image classification.

# Connect to GitHub Classroom

Please follow these steps to connect:

1. Follow this invitation link and accept the invitation: https://classroom.github.com/a/TGh1XJFW
2. The link may ask you to sign in to GitHub (if you haven't signed in earlier). If you don't have a GitHub account, you will need to register.
3. Once you have logged in with GitHub, you may need to select your email address to associate your GitHub account with your email address (if you haven't done it in a previous COMP3420 activity). If you can't find your email address, please skip this step and contact diego.molla-aliod@mq.edu.au so that he can do the association manually.
4. Wait a minute or two, and refresh the browser until it indicates that your assignment repository has been created. Your repository is private to you, and you have administration privileges. Only you and the lecture will have access to it. The repository will be listed under the list of repositories belonging to this offering of COMP3420: https://github.com/orgs/COMP3420-2024S1/repositories
5. In contrast with assignment 1 and the practical sessions, your assignment repository will be empty and will not include starter code. you need to add this Jupyter notebook and commit the changes.

Please use the github repository linked to this GitHub classroom. Make sure that you continuously push commits and you provide useful commit comments. Note the following:

*  **1 mark of the assessment of this assignment is related to good practice with the use of GitHub.**
*  **We will also use github as a tool to check for possible plagiarism or contract cheating. For example, if someone only makes commits on the last day, we may investigate whether there was plagiarism or contract cheating.**


# Tasks

Note from Student (**Referencing**): Throughout the I have used the following resources:

1. ChatGPT (https://chat.openai.com/)
2. Tensorflow Website (https://www.tensorflow.org/)
3. Github Forums for bugs and errors faced, mostly during dependency errors.
4. Stack Overflow for code examples and inspiration
## Task 1 - Data exploration, preparation, and partition (4 marks)

Download the MWD from this site and unzip it:

- https://data.mendeley.com/datasets/4drtyfjtfy/1

You will observe that the zipped file contains 1,125 images representing various weather conditions. To facilitate the assessment of this assignment, please make sure that the images are in a folder named `dataset2` and this folder is in the same place as this jupyter notebook.

### 1.1 - data partition (2 marks)

Generate three CSV files named `my_training.csv`, `my_validation.csv`, and `my_test.csv` that partition the dataset into the training, validation, and test set. Each CSV file contains the following two fields:

- File path
- Image label

For example, the file `my_training.csv` could start like this:

```csv
dataset2/cloudy1.jpg,cloudy
dataset2/shine170.jpg,shine
dataset2/shine116.jpg,shine
```

Make sure that the partitions are created randomly, so that the label distribution is similar in each partition. Also, make sure that the samples are sorted in no particular order (randomly)

Display the label distribution of each partition, and display the first 10 rows of each partition.

The following sample files are available together with these instructions. Your files should look similar to these.

- `training.csv`
- `validation.csv`
- `test.csv`

**For the subsequent tasks in this assignment, use the files we provide (`training.csv`, `validation.csv`, `test.csv`). Do not use the files that you have generated, so that any errors generated by your solution do not carry to the rest of the assignment. Also, the files we provide conveniently removed references to images that have a number of channels different from 3.**




In [1]:
import os
import random
import pandas as pd
import re

# Define paths
dataset_folder = "dataset2"  
output_folder = "./"         
pattern = re.compile(r'\d')

# Ensure dataset folder exists
if not os.path.exists(dataset_folder):
    raise FileNotFoundError(f"The dataset folder '{dataset_folder}' does not exist.")

file_paths = []
labels = []

# Iterate over the files in the dataset folder
for file_name in os.listdir(dataset_folder):
    if file_name.endswith(".jpg"):  
        file_paths.append(os.path.join(dataset_folder, file_name))
        label = pattern.split(file_name)[0]  
        labels.append(label)

# Shuffle the data and labels
combined = list(zip(file_paths, labels))
random.shuffle(combined)
file_paths[:], labels[:] = zip(*combined)

# Partition the dataset
total_samples = len(file_paths)
train_size = int(0.7 * total_samples)
val_size = int(0.15 * total_samples)
test_size = total_samples - train_size - val_size

train_paths = file_paths[:train_size]
train_labels = labels[:train_size]

val_paths = file_paths[train_size:train_size + val_size]
val_labels = labels[train_size:train_size + val_size]

test_paths = file_paths[train_size + val_size:]
test_labels = labels[train_size + val_size:]

# Write to CSV files
def write_to_csv(file_paths, labels, csv_filename):
    df = pd.DataFrame({'File path': file_paths, 'Image label': labels})
    df.to_csv(os.path.join(output_folder, csv_filename), index=False, header = False)

write_to_csv(train_paths, train_labels, 'my_training.csv')
write_to_csv(val_paths, val_labels, 'my_validation.csv')
write_to_csv(test_paths, test_labels, 'my_test.csv')

# Display label distribution and first 10 rows of each partition
def display_partition_info(file_paths, labels, partition_name):
    print(f"{partition_name} partition:")
    label_counts = pd.Series(labels).value_counts()
    print("Label distribution:")
    print(label_counts)
    print("\nFirst 10 rows:")
    print(pd.DataFrame({'File path': file_paths[:10], 'Image label': labels[:10]}))
    print("\n")

display_partition_info(train_paths, train_labels, "Training")
display_partition_info(val_paths, val_labels, "Validation")
display_partition_info(test_paths, test_labels, "Test")


Training partition:
Label distribution:
sunrise    240
cloudy     210
shine      184
rain       151
Name: count, dtype: int64

First 10 rows:
                 File path Image label
0  dataset2\sunrise318.jpg     sunrise
1    dataset2\shine148.jpg       shine
2      dataset2\shine6.jpg       shine
3   dataset2\cloudy270.jpg      cloudy
4   dataset2\cloudy197.jpg      cloudy
5      dataset2\rain63.jpg        rain
6     dataset2\shine77.jpg       shine
7   dataset2\sunrise14.jpg     sunrise
8  dataset2\sunrise250.jpg     sunrise
9     dataset2\shine55.jpg       shine


Validation partition:
Label distribution:
sunrise    64
cloudy     45
shine      31
rain       28
Name: count, dtype: int64

First 10 rows:
                 File path Image label
0  dataset2\sunrise119.jpg     sunrise
1  dataset2\sunrise132.jpg     sunrise
2  dataset2\sunrise278.jpg     sunrise
3   dataset2\sunrise64.jpg     sunrise
4  dataset2\sunrise242.jpg     sunrise
5    dataset2\shine158.jpg       shine
6  dataset2\su

Since the dataset had the name in the following pattern of labelNumber.jpg, to extract the label, I used the number as the pattern for the regular expression. For each file name, I had split the file names from .jpg extension and split the label and number to extract the path of the file and the corresponding label. I had split the datasets at a 70:15:15 ratio for train:validation:test. Using simple python and Pandas functions, I finally created the csv files.

### 1.2 - preprocessing and preparation (2 marks)

Use TensorFlow's `TextLineDataset` to generate datasets for training, validation, and test. The datasets need to produce images that are re-sized to dimensions 230 x 230 and 3 channels, and the values of the pixels must be normalised to the range [0, 1].


In [2]:
import tensorflow as tf

In [3]:
# Function to parse each line of the CSV file
def parse_csv_line(line):
    # Define the format of your data
    columns = tf.io.decode_csv(line, record_defaults=[[""], [""]])
    # Return the file path and label
    return columns[0], columns[1]

# Function to load and preprocess the image
def load_and_preprocess_image(file_path, label):
    # Read the image file
    image = tf.io.read_file(file_path)
    # Decode the image
    image = tf.image.decode_jpeg(image, channels=3)
    # Resize the image to the desired dimensions
    image = tf.image.resize(image, [230, 230])
    # Normalize the pixel values to the range [0, 1]
    image = tf.cast(image, tf.float32) / 255.0
    return image, label

In [4]:
IMG_HEIGHT = 230
IMG_WIDTH = 230

# Paths to your CSV files
train_file = "training.csv"
val_file = "validation.csv"
test_file = "test.csv"

# Load and preprocess the datasets
train_dataset = tf.data.TextLineDataset(train_file).map(parse_csv_line).map(load_and_preprocess_image)
val_dataset = tf.data.TextLineDataset(val_file).map(parse_csv_line).map(load_and_preprocess_image)
test_dataset = tf.data.TextLineDataset(test_file).map(parse_csv_line).map(load_and_preprocess_image)


# Display dataset element specifications
print("Train Dataset Spec:", train_dataset.element_spec)
print("Validation Dataset Spec:", val_dataset.element_spec)
print("Test Dataset Spec:", test_dataset.element_spec)


Train Dataset Spec: (TensorSpec(shape=(230, 230, 3), dtype=tf.float32, name=None), TensorSpec(shape=(), dtype=tf.string, name=None))
Validation Dataset Spec: (TensorSpec(shape=(230, 230, 3), dtype=tf.float32, name=None), TensorSpec(shape=(), dtype=tf.string, name=None))
Test Dataset Spec: (TensorSpec(shape=(230, 230, 3), dtype=tf.float32, name=None), TensorSpec(shape=(), dtype=tf.string, name=None))


I have created a preprocessing function named `load_and_preprocess_image` takes an image path and its corresponding label as input. It reads the image file, decodes it, ensures it has 3 channels (RGB), resizes it to 230x230 dimensions, and normalizes the pixel values to the range [0, 1]. It then returns the preprocessed image and its label. I tried implementing as much abstraction as possible throughout the assignment. 

The function named load_and_preprocess_image which `loads datasets` using TensorFlow's `TextLineDataset`. It takes a CSV file path as input. Inside the function, each line of the CSV file is parsed to extract the image path and its label using `tf.io.decode_csv`. Then, the `preprocess_image` function is applied to each element of the dataset to preprocess the images. Parsing is necessary for successful and correct data extraction. The delimiter makes sure that the correct label and corresponding path is being parsed for the `TextLineDataset`.

#### Loading and Preprocessing Images:

- This function loads and preprocesses an image given its file path and corresponding label.

- It reads the image file using tf.io.read_file.

- It decodes the JPEG-encoded image into a tensor using tf.image.decode_jpeg.

- It resizes the image to the desired dimensions of 230x230 pixels using tf.image.resize.

- It normalizes the pixel values to the range [0, 1] by casting to tf.float32 and dividing by 255.0.
    
- Finally, it returns the preprocessed image tensor along with its label.

These loops above iterate over the first two elements of each dataset (`train_dataset`, `val_dataset`, `test_dataset`) and print the shape of the image tensor and its corresponding label. This is just for demonstration purposes that the preprocessing and dataset creation has been performed successfully.

## Task 2 - A simple classifier (4 marks)


The following `X` and `Y` values for test, validation and training will be used for the later codes below until Task 3.1.2 when it requires an image shape of `(224,224,3)`. A new function, `load_and_preprocess_dataset` was made to adhere to the specification of labels.

In [6]:
IMG_HEIGHT = 230
IMG_WIDTH = 230

X_train, y_train = load_and_preprocess_dataset("training.csv")
X_val, y_val = load_and_preprocess_dataset("validation.csv")
X_test, y_test = load_and_preprocess_dataset("test.csv")


In [5]:
import pandas as pd
import numpy as np

def load_and_preprocess_dataset(file_path):
    # Load the CSV file
    df = pd.read_csv(file_path, header=None, names=['File', 'Label'])
    
    # Shuffle the dataset
    df = df.sample(frac=1).reset_index(drop=True)
    
    # Separate file paths and labels
    file_paths = df['File'].values
    labels = df['Label'].values
    
    # Preprocess file paths to load images
    images = []
    for file_path in file_paths:
        # Load and preprocess image (e.g., resizing, normalizing)
        img = tf.keras.preprocessing.image.load_img(file_path, target_size=(IMG_HEIGHT, IMG_WIDTH))
        img = tf.image.convert_image_dtype(img, tf.float32)
        images.append(img)
    
    # Convert labels to numeric
    label_to_numeric = {'rain': 0, 'sunrise': 1, 'cloudy': 2, 'shine': 3}
    numeric_labels = np.array([label_to_numeric[label] for label in labels])
    
    return np.array(images), numeric_labels



### 2.1 First classifier (1 mark)

Create a simple model that contains the following layers:

- A `Flatten` layer.
- The output layer with the correct size and activation function for this classification task.

Then, train the model with the training data. Use the validation data to determine when to stop training. Finally, test the trained model on the test data and report the accuracy.

In [7]:
import keras as keras 

# Define input shape and number of classes
input_shape = (IMG_HEIGHT, IMG_WIDTH, 3)
num_classes = 4  # Assuming you have 4 classes: rain, sunrise, cloudy, shine

def create_simple_model(input_shape, num_classes):
    model = tf.keras.models.Sequential([
        tf.keras.layers.Flatten(input_shape=input_shape),
        tf.keras.layers.Dense(num_classes, activation='softmax')
    ])
    return model

# Create the simple model
simple_model = create_simple_model(input_shape, num_classes)

# Compile the model
simple_model.compile(optimizer='adam',
                     loss='sparse_categorical_crossentropy',
                     metrics=['accuracy'])

# Train the model
history = simple_model.fit(X_train, y_train,
                           epochs=10,
                           validation_data=(X_val, y_val),
                           callbacks=[keras.callbacks.EarlyStopping(patience=3)])

# Evaluate the model on test data
test_loss, test_accuracy = simple_model.evaluate(X_test, y_test)
print("Test Accuracy:", test_accuracy)


Epoch 1/10


25/25 [==============================] - 1s 26ms/step - loss: 32.1647 - accuracy: 0.4750 - val_loss: 10.4859 - val_accuracy: 0.6347
Epoch 2/10
25/25 [==============================] - 0s 19ms/step - loss: 6.7026 - accuracy: 0.6786 - val_loss: 4.5981 - val_accuracy: 0.7425
Epoch 3/10
25/25 [==============================] - 0s 20ms/step - loss: 4.8364 - accuracy: 0.7362 - val_loss: 3.7194 - val_accuracy: 0.7425
Epoch 4/10
25/25 [==============================] - 0s 19ms/step - loss: 3.9308 - accuracy: 0.7606 - val_loss: 9.6176 - val_accuracy: 0.6347
Epoch 5/10
25/25 [==============================] - 0s 19ms/step - loss: 4.0354 - accuracy: 0.7618 - val_loss: 2.9328 - val_accuracy: 0.7605
Epoch 6/10
25/25 [==============================] - 1s 21ms/step - loss: 2.2589 - accuracy: 0.7951 - val_loss: 3.9397 - val_accuracy: 0.6587
Epoch 7/10
25/25 [==============================] - 0s 19ms/step - loss: 3.1572 - accuracy: 0.7529 - val_loss: 2.3107 - val_accuracy: 0.7665
Epoch 8/10
25/25 [====

### Model Architecture:

For the simple classifier, I chose a basic architecture consisting of a single Flatten layer followed by a Dense output layer with softmax activation. The Flatten layer is used to flatten the input images into a one-dimensional tensor, while the Dense layer with softmax activation produces probabilities for each class.

Training:

- Optimizer: Adam optimizer is used for training due to its adaptive learning rate capabilities and efficiency in handling large datasets.
- Loss Function: Sparse categorical crossentropy is chosen as the loss function since it's suitable for multi-class classification tasks where the labels are integers.
- Metrics: Accuracy is selected as the evaluation metric to monitor the model's performance during training.

Early stopping is implemented with a patience of 3 epochs to prevent overfitting. It monitors the validation loss and stops training if there's no improvement for consecutive epochs, thereby preventing the model from learning noise in the training data.

### 2.2 A more complex classifier (2 marks)

Try a more complex architecture that has 1 or more hidden layers with dropout. For this more complex architecture, use `keras-tuner` and run it with a reasonable choice of possible parameters. You may try among the following:

- Number of hidden layers
- Sizes of hidden layers
- Dropout rate
- Learning rate

In [8]:
import tensorflow as tf
import kerastuner as kt

# Define a function to build the model
def build_model(hp):
    model = tf.keras.models.Sequential()
    model.add(tf.keras.layers.Flatten(input_shape=(230, 230, 3)))
    
    # Add one or more hidden layers with dropout
    for i in range(hp.Int('num_layers', 1, 3)):
        model.add(tf.keras.layers.Dense(
            units=hp.Int(f'units_{i}', min_value=32, max_value=512, step=32),
            activation='relu'))
        model.add(tf.keras.layers.Dropout(
            rate=hp.Float(f'dropout_{i}', min_value=0.1, max_value=0.5, step=0.1)))
    
    model.add(tf.keras.layers.Dense(4, activation='softmax'))  # Assuming 4 classes
    
    # Compile the model
    model.compile(optimizer=tf.keras.optimizers.Adam(
        hp.Choice('learning_rate', values=[1e-2, 1e-3, 1e-4])),
        loss='sparse_categorical_crossentropy',
        metrics=['accuracy'])
    
    return model

# Instantiate the tuner
tuner = kt.Hyperband(
    build_model,
    objective='val_accuracy',
    max_epochs=10,
    directory='my_dir',
    project_name='more_complex_classifier')

# Perform the hyperparameter search
tuner.search(X_train, y_train,
             epochs=10,
             validation_data=(X_val, y_val))

# Get the best hyperparameters
best_hps = tuner.get_best_hyperparameters(num_trials=1)[0]

# Build the model with the optimal hyperparameters
model = tuner.hypermodel.build(best_hps)

# Train the model
history = model.fit(X_train, y_train,
                    epochs=10,
                    validation_data=(X_val, y_val),
                    callbacks=[tf.keras.callbacks.EarlyStopping(patience=3)])

# Evaluate the model on test data
test_loss, test_accuracy = model.evaluate(X_test, y_test)
print("Test Accuracy:", test_accuracy)
print("Best Hyperparameters:", best_hps.values)




C:\Users\rakee\AppData\Local\Temp\ipykernel_30664\2594156545.py:2: DeprecationWarning: `import kerastuner` is deprecated, please use `import keras_tuner`.
  import kerastuner as kt


Reloading Tuner from my_dir\more_complex_classifier\tuner0.json
Epoch 1/10
25/25 [==============================] - 18s 670ms/step - loss: 4.8343 - accuracy: 0.5198 - val_loss: 0.9731 - val_accuracy: 0.7305
Epoch 2/10
25/25 [==============================] - 16s 641ms/step - loss: 1.0911 - accuracy: 0.6530 - val_loss: 0.7592 - val_accuracy: 0.6647
Epoch 3/10
25/25 [==============================] - 17s 671ms/step - loss: 0.6920 - accuracy: 0.7068 - val_loss: 0.7012 - val_accuracy: 0.7365
Epoch 4/10
25/25 [==============================] - 16s 651ms/step - loss: 0.6435 - accuracy: 0.7426 - val_loss: 0.7021 - val_accuracy: 0.7246
Epoch 5/10
25/25 [==============================] - 17s 678ms/step - loss: 0.6537 - accuracy: 0.7260 - val_loss: 0.6576 - val_accuracy: 0.7665
Epoch 6/10
25/25 [==============================] - 17s 675ms/step - loss: 0.6699 - accuracy: 0.7119 - val_loss: 0.7583 - val_accuracy: 0.7066
Epoch 7/10
25/25 [==============================] - 17s 677ms/step - loss: 0.6

Write text below where you explain and justify your decision choices made in this task.

I created the function create_complex_model which defines the architecture of a neural network model.
- The input shape of the images is specified as `(230, 230, 3)`, representing images with height and width of 230 pixels and 3 color channels (RGB).
The model consists is made with a helper function which builds the model using the best hyperparameters as follows: 
- `'num_layers'`: `1` specifies that the best model has 1 hidden layer.
- `'units_0'`: `416` indicates the number of units (neurons) in the first hidden layer.
- `'dropout_0'`: `0.4` specifies the dropout rate applied to the first hidden layer.
- `'learning_rate'`: `0.0001` indicates the learning rate used for the Adam optimizer.
- `'units_1'`: `352` specifies the number of units in the second hidden layer (if applicable).
- `'dropout_1'`: `0.3` indicates the dropout rate applied to the second hidden layer (if applicable).
- `'units_2'`: `160` specifies the number of units in the third hidden layer (if applicable).
- `'dropout_2'`: `0.1` indicates the dropout rate applied to the third hidden layer (if applicable).
- `'tuner/epochs'`: `10` specifies the maximum number of epochs used during the hyperparameter search.
- `'tuner/initial_epoch'`: `4` indicates the initial epoch used during the hyperparameter search.
- `'tuner/bracket'`: `1` specifies the hyperband bracket used during the hyperparameter search.
- `'tuner/round'`: `1` specifies the round of the hyperband algorithm used during the hyperparameter search.
- `'tuner/trial_id'`: `'0014'` uniquely identifies the trial that produced these hyperparameters.

The Test accuracy of the model comes out to be `0.822` which is higher than the test accuracy of the simple model at `0.787`. Due to additional layers with the best hyperparameter search, we can find the best model accuracy and accuracy of the simple model was much lower than the test accuracy of the complex model accuracy. This was due to higher number of layers has better accuracy but takes more time to complete.

### 2.3 Error analysis (1 mark)

Evaluate your best-performing system from task 2 against the system of task 1 and answer the following questions.

1. Which system had a better accuracy on the test data?
2. Which system had a lower degree of overfitting?

1. From the above outputs we can see that the `complex` model had a higher accuracy on the test data than the simpel. And so system at `task 2` had a higher accuracy on the test data.
2. System at `task 1` had a `lower degree of overfitting` and this can be seen from the lesser number of hidden layers. More complex systems tend to have a `higher degree of accuracy` but comes with the `higher degree of overfitting`.

## Task 3 - A more complex classifier (5 marks)

### Task 3.1 Using ConvNets (2 marks)

Implement a model that uses a sequence of at least two `ConvD`, each one followed with `MaxPooling2D`. Use reasonable numbers for the hyperparameters (number of filters, kernel size, pool size, activation, etc), base on what we have seen in the lectures. Feel free to research the internet and / or generative AI to help you find a reasonable choice of hyperparameters. For this task, do not use pre-trained models.

In [9]:
from sklearn.metrics import classification_report

input_shape = (IMG_HEIGHT, IMG_WIDTH, 3)

def create_convnet_model(input_shape, num_classes):
    model = tf.keras.Sequential([
        tf.keras.layers.Conv2D(32, (3, 3), activation='relu', input_shape=input_shape),
        tf.keras.layers.MaxPooling2D((2, 2)),
        tf.keras.layers.Conv2D(64, (3, 3), activation='relu'),
        tf.keras.layers.MaxPooling2D((2, 2)),
        tf.keras.layers.Flatten(input_shape=input_shape),
        tf.keras.layers.Dense(128, activation='relu'),
        tf.keras.layers.Dense(num_classes, activation='softmax')
    ])
    return model

# Define input shape and number of classes
input_shape = (230, 230, 3)  # Input images are resized to (230, 230) with 3 channels
num_classes = 4  # 4 classes: rain, sunrise, cloudy, shine

# Create the ConvNet model
convnet_model = create_convnet_model(input_shape, num_classes)

# Compile the model
convnet_model.compile(optimizer='adam',
                      loss='sparse_categorical_crossentropy',
                      metrics=['accuracy'])

# Display the model summary
convnet_model.summary()

# Train the model
history = convnet_model.fit(X_train, y_train,
                    epochs=15,
                    validation_data=(X_val, y_val),
                    callbacks=[tf.keras.callbacks.EarlyStopping(patience=3)])

# Evaluate the model
test_loss, test_accuracy = model.evaluate(X_test, y_test)
print("Test Loss:", test_loss)
print("Test Accuracy:", test_accuracy)

y_pred_probs = convnet_model.predict(X_test)

# Convert predicted probabilities to class labels
y_pred = np.argmax(y_pred_probs, axis=1)

# Generate classification report
class_names = ['rain', 'sunrise', 'cloudy', 'shine']
print(classification_report(y_test, y_pred, target_names=class_names))


Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 228, 228, 32)      896       
                                                                 
 max_pooling2d (MaxPooling2  (None, 114, 114, 32)      0         
 D)                                                              
                                                                 
 conv2d_1 (Conv2D)           (None, 112, 112, 64)      18496     
                                                                 
 max_pooling2d_1 (MaxPoolin  (None, 56, 56, 64)        0         
 g2D)                                                            
                                                                 
 flatten_2 (Flatten)         (None, 200704)            0         
                                                                 
 dense_3 (Dense)             (None, 128)              

Epoch 1/15
25/25 [==============================] - 24s 924ms/step - loss: 3.5236 - accuracy: 0.5288 - val_loss: 0.6864 - val_accuracy: 0.6587
Epoch 2/15
25/25 [==============================] - 23s 943ms/step - loss: 0.5578 - accuracy: 0.7875 - val_loss: 0.5715 - val_accuracy: 0.8443
Epoch 3/15
25/25 [==============================] - 23s 923ms/step - loss: 0.4268 - accuracy: 0.8566 - val_loss: 0.5416 - val_accuracy: 0.8802
Epoch 4/15
25/25 [==============================] - 22s 890ms/step - loss: 0.2847 - accuracy: 0.9027 - val_loss: 0.5517 - val_accuracy: 0.8922
Epoch 5/15
25/25 [==============================] - 22s 895ms/step - loss: 0.2085 - accuracy: 0.9296 - val_loss: 0.4262 - val_accuracy: 0.9042
Epoch 6/15
25/25 [==============================] - 23s 920ms/step - loss: 0.1502 - accuracy: 0.9398 - val_loss: 0.4729 - val_accuracy: 0.8922
Epoch 7/15
25/25 [==============================] - 23s 903ms/step - loss: 0.1177 - accuracy: 0.9616 - val_loss: 0.4170 - val_accuracy: 0.8922

### Task 3.2 Using pre-trained models (2 marks)

Use MobileNet, pre-trained on imagenet as discussed in the lectures. Add the correct classification layer, and train it with your data. Make sure that you freeze MobileNet's weights during training. Also, make sure you use a reasonable schedule for the learning rate.

In [10]:
import os
IMG_HEIGHT = 224
IMG_WIDTH = 224
IMG_CHANNELS = 3

X_train, y_train = load_and_preprocess_dataset("training.csv")
X_val, y_val = load_and_preprocess_dataset("validation.csv")
X_test, y_test = load_and_preprocess_dataset("test.csv")

In [11]:
# Load compressed models from tensorflow_hub
os.environ['TFHUB_MODEL_LOAD_FORMAT'] = 'COMPRESSED'

import tensorflow as tf
import tensorflow_hub as hub

input_shape = (IMG_HEIGHT, IMG_WIDTH, IMG_CHANNELS)

# Define the model architecture
model = tf.keras.Sequential([hub.KerasLayer("https://tfhub.dev/google/tf2-preview/mobilenet_v2/feature_vector/4",
                                            input_shape=input_shape, trainable=False, name='mobilenet_embedding'),
                            tf.keras.layers.Dense(16,activation='relu', name='dense_hidden'),
                            tf.keras.layers.Dense(4,activation='softmax',name='weather_prob')])


# Compile the model
model.compile(optimizer=tf.keras.optimizers.Adam(learning_rate=0.001),
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

# Display model summary
print(model.summary())

# Train the model
history = model.fit(X_train, y_train,
                    epochs=15,
                    validation_data=(X_val, y_val),
                    callbacks=[tf.keras.callbacks.EarlyStopping(patience=3)])

# Evaluate the model
test_loss, test_accuracy = model.evaluate(X_test, y_test)
print("Test Loss:", test_loss)
print("Test Accuracy:", test_accuracy)

y_pred_probs = model.predict(X_test)

# Convert predicted probabilities to class labels
y_pred = np.argmax(y_pred_probs, axis=1)

# Generate classification report
class_names = ['rain', 'sunrise', 'cloudy', 'shine']
print(classification_report(y_test, y_pred, target_names=class_names))

Model: "sequential_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 mobilenet_embedding (Keras  (None, 1280)              2257984   
 Layer)                                                          
                                                                 
 dense_hidden (Dense)        (None, 16)                20496     
                                                                 
 weather_prob (Dense)        (None, 4)                 68        
                                                                 
Total params: 2278548 (8.69 MB)
Trainable params: 20564 (80.33 KB)
Non-trainable params: 2257984 (8.61 MB)
_________________________________________________________________
None
Epoch 1/15
25/25 [==============================] - 13s 416ms/step - loss: 0.8804 - accuracy: 0.6415 - val_loss: 0.3924 - val_accuracy: 0.8683
Epoch 2/15
25/25 [==============================] - 10s 392ms/s

### Task 3.3 Comparative evaluation (1 mark)

Compare the evaluation results of the best systems from tasks 3.1 and 3.2 and answer the following questions.

1. What system (including the systems you developed in Task 2) perform best on the test set?
2. Report the accuracy of your best system on each of the different weather categories. What type of weather was most difficult to detect?

1. From the above outputs we can see that the `pre-trained` model had a higher accuracy, of `0.964` on the test data than just the ConvNet one. And so system at `task 2` had a higher accuracy on the test data.

2. **System 1**: 
    |         | Precision | Recall | F1-score | Support |
    |---------|-----------|--------|----------|---------|
    | rain    |    0.94   |  0.97  |   0.96   |    34   |
    | sunrise |    1.00   |  0.98  |   0.99   |    49   |
    | cloudy  |    0.84   |  0.94  |   0.89   |    51   |
    | shine   |    0.90   |  0.74  |   0.81   |    35   |
    
    **System 2**:
    |          | Precision  | Recall  | F1-score  | Support  |              
    |:---------|-----------:|--------:|----------:|---------:|
    | rain     |    1.00    |  1.00   |   1.00    |    34    |
    | sunrise  |    0.98    |  0.98   |   0.98    |    49    |
    | cloudy   |    0.98    |  0.92   |   0.95    |    51    |
    | shine    |    0.92    |  1.00   |   0.96    |    35    |              

From the above tables extracted from the output of the code, we can analyse the difficulty level using the F1-score. Based on the F1-scores, we can see that for System 1, "shine" has the lowest F1-score of 0.81, making it the most difficult weather category to detect for System 1. Similarly, for System 2, "cloudy" has the lowest F1-score of 0.95, making it the most difficult weather category to detect for System 2.



## Coding (1 mark)

This mark will be assigned to submissions that have clean and efficient code and good in-code documentation of all code presented in this assignment.

## GitHub Classroom (1 mark)

These marks will be given to submissions that:

- Have continuously committed changes to the GitHub repository at GitHub Classroom.
- The commit messages are useful and informative.

# Submission

Your submission should consist of this Jupyter notebook with all your code and explanations inserted into the notebook as text cells. **The notebook should contain the output of the runs. All code should run. Code with syntax errors or code without output will not be assessed.**

**Do not submit multiple files. If you feel you need to submit multiple files, please contact Diego.Molla-Aliod@mq.edu.au first.**

Examine the text cells of this notebook so that you can have an idea of how to format text for good visual impact. You can also read this useful [guide to the MarkDown notation](https://daringfireball.net/projects/markdown/syntax), which explains the format of the text cells.

Each task specifies a number of marks. The final mark of the assignment is the sum of all the marks of each individual task.

By submitting this assignment you are acknowledging that this is your own work. Any submissions that break the code of academic honesty will be penalised as per [the academic integrity policy](https://policies.mq.edu.au/document/view.php?id=3).

## A note on the use of AI code generators

In this assignment, we view AI code generators such as copilot, CodeGPT, etc as tools that can help you write code quickly. You are allowed to use these tools, but with some conditions. To understand what you can and what you cannot do, please visit these information pages provided by Macquarie University.

- Artificial Intelligence Tools and Academic Integrity in FSE - https://bit.ly/3uxgQP4

If you choose to use these tools, make the following explicit in your Jupyter notebook, under a section with heading "Use of AI generators in this assignment" :

- What part of your code is based on the output of such tools,
- What tools you used,
- What prompts you used to generate the code or text, and
- What modifications you made on the generated code or text.

This will help us assess your work fairly.
